File: backend/app/api/attendance.py
"""Manual attendance management in the authenticated kindergarten."""

from datetime import date
from typing import Annotated, Literal
from uuid import UUID

from fastapi import APIRouter, Depends, Query, Response
from sqlalchemy.orm import Session

from app.api.error_responses import MANAGEMENT_ERRORS
from app.core.permissions import require_role
from app.db.session import get_db
from app.models.user import User
from app.schemas.attendance import (
    AttendanceCreate,
    AttendanceDetail,
    AttendanceList,
    AttendancePatch,
)
from app.services import attendance

router = APIRouter(prefix="/attendance", tags=["Посещаемость"], responses=MANAGEMENT_ERRORS)
Manager = Annotated[User, Depends(require_role("DIRECTOR", "ADMIN"))]
Database = Annotated[Session, Depends(get_db)]


@router.get("", response_model=AttendanceList)
def list_attendance(user: Manager, db: Database, date: date,
                    group_id: UUID | None = None, child_id: UUID | None = None,
                    status: Literal["all", "present", "absent", "unknown"] = Query("all")) -> AttendanceList:
    return AttendanceList(items=attendance.list_day(db, user, date, group_id, status, child_id))


@router.post("", response_model=AttendanceDetail, status_code=201)
def save_attendance(payload: AttendanceCreate, user: Manager, db: Database, response: Response) -> AttendanceDetail:
    result, created = attendance.upsert(db, user, payload)
    if not created:
        response.status_code = 200
    return result


@router.get("/{record_id}", response_model=AttendanceDetail)
def get_attendance(record_id: UUID, user: Manager, db: Database) -> AttendanceDetail:
    return attendance.get(db, user, record_id)


@router.patch("/{record_id}", response_model=AttendanceDetail)
def patch_attendance(record_id: UUID, payload: AttendancePatch, user: Manager, db: Database) -> AttendanceDetail:
    return attendance.update(db, user, record_id, payload)

File: frontend/src/lib/api/attendance.ts
import { api } from "./client";
import type { AttendanceDetail, AttendanceRow, AttendanceStatus } from "@/types/stage3";

export const attendanceApi = {
  list: (date: string, groupId?: string, status?: AttendanceStatus | "all", childId?: string) => {
    const query = new URLSearchParams({ date });
    if (groupId) query.set("group_id", groupId);
    if (status && status !== "all") query.set("status", status);
    if (childId) query.set("child_id", childId);
    return api.get<{ items: AttendanceRow[] }>(`/attendance?${query}`);
  },
  save: (body: { child_id: string; date: string; status: AttendanceStatus; arrival_time: string | null; departure_time: string | null }) => api.post<AttendanceDetail>("/attendance", body),
  get: (id: string) => api.get<AttendanceDetail>(`/attendance/${encodeURIComponent(id)}`),
  patch: (id: string, body: Partial<Pick<AttendanceDetail, "status" | "arrival_time" | "departure_time">>) => api.patch<AttendanceDetail>(`/attendance/${encodeURIComponent(id)}`, body),
};
